Terraform crash upon `azurerm_private_dns_zone` destruction

_This issue was originally opened by @esciara as hashicorp/terraform#28322. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._
    
<hr>
    
<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers.

To fix problems, we need clear reproduction cases - we need to be able to see it happen locally. A reproduction case is ideally something a Terraform Core engineer can git-clone or copy-paste and run immediately, without inventing any details or context. 

* A short example can be directly copy-pasteable; longer examples should be in separate git repositories, especially if multiple files are needed
* Please include all needed context. For example, if you figured out that an expression can cause a crash, put the expression in a variable definition or a resource
* Set defaults on (or omit) any variables. The person reproducing it should not need to invent variable settings
* If multiple steps are required, such as running terraform twice, consider scripting it in a simple shell script. For example, see [this case](https://github.com/danieldreier/terraform-issue-reproductions/tree/master/25719). Providing a script can be easier than explaining what changes to make to the config between runs.
* Omit any unneeded complexity: remove variables, conditional statements, functions, modules, providers, and resources that are not needed to trigger the bug
* When possible, use the [null resource](https://www.terraform.io/docs/providers/null/resource.html) provider rather than a real provider in order to minimize external dependencies. We know this isn't always feasible. The Terraform Core team doesn't have deep domain knowledge in every provider, or access to every cloud platform for reproduction cases.

-->

### Terraform Version

```
2021/04/09 16:56:23 [INFO] Terraform version: 0.14.10
2021/04/09 16:56:23 [INFO] Go runtime version: go1.16.3
2021/04/09 16:56:23 [INFO] CLI args: []string{"/usr/local/bin/terraform", "apply"}

2021-04-09T16:56:24.304+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For Terraform configs larger than a few resources, or that involve multiple files, please make a GitHub repository that we can clone, rather than copy-pasting multiple files in here. For security, you can also encrypt the files using our GPG public key at https://www.hashicorp.com/security.
-->

```terraform
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "test" {
  location = "France Central"
  name     = "test-resource-group"
}

resource "azurerm_private_dns_zone" "test" {
  name                = "test.cloud"
  resource_group_name = azurerm_resource_group.test.name
}

resource "azurerm_virtual_network" "test" {
  name                = "test-virtual-network"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  address_space       = ["172.24.0.0/20"]
}

resource "azurerm_private_dns_zone_virtual_network_link" "test" {
  name                  = "test-dns-to-vnet-link"
  private_dns_zone_name = "test.cloud"
  resource_group_name   = azurerm_resource_group.test.name
  virtual_network_id    = azurerm_virtual_network.test.id
}
```

### Debug Output

See `crash.log` file.

### Crash Output

Need some support to ensure that there is no sensitive information left in the editored `crash.log` file before sharing it.

### Expected Behavior

When removing the `azurerm_private_dns_zone` resource from the terraform configuration file after the first `apply` that created it, the subsequent `apply` should not succeed and terraform inform us gracefully about it.

### Actual Behavior
Terraform crashes with the following output:

```
azurerm_private_dns_zone.test: Destroying... [id=/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/test-resource-group/providers/Microsoft.Network/privateDnsZones/test.cloud]

Error: rpc error: code = Unavailable desc = transport is closing


panic: runtime error: invalid memory address or nil pointer dereference
2021-04-09T16:56:51.469+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5: [signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x5776fe4]
2021-04-09T16:56:51.469+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5: 
2021-04-09T16:56:51.469+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5: goroutine 114 [running]:
2021-04-09T16:56:51.469+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5: github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/privatedns.resourcePrivateDnsZoneDelete(0xc001410a80, 0x5d59ea0, 0xc000340300, 0x0, 0x0)
2021-04-09T16:56:51.469+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/privatedns/private_dns_zone_resource.go:270 +0x244
2021-04-09T16:56:51.469+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).Apply(0xc000b250e0, 0xc000296a50, 0xc001adeaa0, 0x5d59ea0, 0xc000340300, 0x100df01, 0xc001790a20, 0x6110020)
2021-04-09T16:56:51.470+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/helper/schema/resource.go:283 +0x4a6
2021-04-09T16:56:51.470+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).Apply(0xc000123c80, 0xc000555a38, 0xc000296a50, 0xc001adeaa0, 0x9e92ee0, 0x0, 0x0)
2021-04-09T16:56:51.470+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/helper/schema/provider.go:294 +0x99
2021-04-09T16:56:51.470+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5: github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ApplyResourceChange(0xc000288e48, 0x6be2750, 0xc001051e90, 0xc001410460, 0xc000288e48, 0xc001051e90, 0xc0007cdba0)
2021-04-09T16:56:51.470+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin/grpc_provider.go:885 +0x8a5
2021-04-09T16:56:51.470+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5: github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x622ede0, 0xc000288e48, 0x6be2750, 0xc001051e90, 0xc00074c420, 0x0, 0x6be2750, 0xc001051e90, 0xc001320c00, 0x2dd)
2021-04-09T16:56:51.470+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5/tfplugin5.pb.go:3305 +0x214
2021-04-09T16:56:51.470+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5: google.golang.org/grpc.(*Server).processUnaryRPC(0xc000224e00, 0x6c25618, 0xc0005ca480, 0xc0015f0200, 0xc0006969c0, 0x9e50720, 0x0, 0x0, 0x0)
2021-04-09T16:56:51.470+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:1180 +0x52b
2021-04-09T16:56:51.470+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5: google.golang.org/grpc.(*Server).handleStream(0xc000224e00, 0x6c25618, 0xc0005ca480, 0xc0015f0200, 0x0)
2021-04-09T16:56:51.470+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:1503 +0xd0c
2021-04-09T16:56:51.470+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5: google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000d7c220, 0xc000224e00, 0x6c25618, 0xc0005ca480, 0xc0015f0200)
2021-04-09T16:56:51.470+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:843 +0xab
2021-04-09T16:56:51.470+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5: created by google.golang.org/grpc.(*Server).serveStreams.func1
2021-04-09T16:56:51.470+0200 [DEBUG] plugin.terraform-provider-azurerm_v2.55.0_x5:      /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:841 +0x1fd
2021-04-09T16:56:51.477+0200 [WARN]  plugin.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2021-04-09T16:56:51.477+0200 [DEBUG] plugin: plugin process exited: path=.terraform/providers/registry.terraform.io/hashicorp/azurerm/2.55.0/darwin_amd64/terraform-provider-azurerm_v2.55.0_x5 pid=38570 error="exit status 2"
2021/04/09 16:56:51 [DEBUG] azurerm_private_dns_zone.test: apply errored, but we're indicating that via the Error pointer rather than returning it: rpc error: code = Unavailable desc = transport is closing
2021/04/09 16:56:51 [TRACE] EvalWriteState: writing current state object for azurerm_private_dns_zone.test
2021/04/09 16:56:51 [TRACE] vertex "azurerm_private_dns_zone.test (destroy)": visit complete
2021/04/09 16:56:51 [TRACE] dag/walk: upstream of "provider[\"registry.terraform.io/hashicorp/azurerm\"] (close)" errored, so skipping
2021/04/09 16:56:51 [TRACE] dag/walk: upstream of "meta.count-boundary (EachMode fixup)" errored, so skipping
2021/04/09 16:56:51 [TRACE] dag/walk: upstream of "root" errored, so skipping
2021/04/09 16:56:51 [TRACE] statemgr.Filesystem: creating backup snapshot at terraform.tfstate.backup
2021/04/09 16:56:51 [TRACE] statemgr.Filesystem: state has changed since last snapshot, so incrementing serial to 6
2021/04/09 16:56:51 [TRACE] statemgr.Filesystem: writing snapshot at terraform.tfstate
2021/04/09 16:56:51 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info
2021/04/09 16:56:51 [TRACE] statemgr.Filesystem: unlocking terraform.tfstate using fcntl flock
2021-04-09T16:56:51.492+0200 [DEBUG] plugin: plugin exited



!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

SECURITY WARNING: the "crash.log" file that was created may contain 
sensitive information that must be redacted before it is safe to share 
on the issue tracker.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
### Steps to Reproduce
1. `terraform init`
2. `terraform apply`
3. Remove the `azurerm_private_dns_zone` resource from the configuration file
4. `terraform apply` => crash

### Additional Context
N/A

### References
N/A

Seems like this happens for any Azure resource that has a delete lock on it.

```
Stack trace from the terraform-provider-azurerm_v2.61.0_x5 plugin:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x4d27707]
goroutine 100 [running]:
github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/postgres.resourcePostgreSQLFirewallRuleDelete(0xc000860cb0, 0x53ff0a0, 0xc0003be900, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/postgres/postgresql_firewall_rule_resource.go:174 +0x247
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).Apply(0xc000a166e0, 0xc00139d3b0, 0xc001a76320, 0x53ff0a0, 0xc0003be900, 0x40e101, 0xc0000e5200, 0x57898c0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/helper/schema/resource.go:293 +0x4a6
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).Apply(0xc000195900, 0xc0013a1a38, 0xc00139d3b0, 0xc001a76320, 0x97214e0, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/helper/schema/provider.go:294 +0x99
github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ApplyResourceChange(0xc0003065b0, 0x62e48b0, 0xc0017208d0, 0xc000860620, 0xc0003065b0, 0xc0017208d0, 0xc000e79ba0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin/grpc_provider.go:895 +0x8a5
github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x58e16a0, 0xc0003065b0, 0x62e48b0, 0xc0017208d0, 0xc001db4480, 0x0, 0x62e48b0, 0xc0017208d0, 0xc0006fd680, 0x21b)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5/tfplugin5.pb.go:3305 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001ff340, 0x632a058, 0xc000102900, 0xc001746300, 0xc0002d24e0, 0x96de900, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:1194 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc0001ff340, 0x632a058, 0xc000102900, 0xc001746300, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:1517 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000d6ea40, 0xc0001ff340, 0x632a058, 0xc000102900, 0xc001746300)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:859 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:857 +0x1fd
Error: The terraform-provider-azurerm_v2.61.0_x5 plugin crashed!
```

--- a/azurerm/internal/services/privatedns/private_dns_zone_resource.go
+++ b/azurerm/internal/services/privatedns/private_dns_zone_resource.go
@@ -6,7 +6,6 @@ import (
 	"time"
 
 	"github.com/Azure/azure-sdk-for-go/services/privatedns/mgmt/2018-09-01/privatedns"
-	"github.com/hashicorp/go-azure-helpers/response"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/azure"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/helpers/tf"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/clients"
@@ -266,17 +265,11 @@ func resourcePrivateDnsZoneDelete(d *pluginsdk.ResourceData, meta interface{}) e
 	etag := ""
 	future, err := client.Delete(ctx, id.ResourceGroup, id.Name, etag)
 	if err != nil {
-		if response.WasNotFound(future.Response()) {
-			return nil
-		}
 		return fmt.Errorf("error deleting Private DNS Zone %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
 	}
 
 	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
-		if response.WasNotFound(future.Response()) {
-			return nil
-		}
-		return fmt.Errorf("error deleting Private DNS Zone %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
+		return fmt.Errorf("waiting for deletion of Private DNS Zone %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
 	}
 
 	return nil

